### ISSUE TITLE
[Question] Loss difference between TransformerWrapper & AutoregressiveWrapper

### ISSUE THREAD

**Issue by pradeep-pyro** [lucidrains/x-transformers#297]
Hi @lucidrains,

For an autoregressive Transformer, if I use `TransformerWrapper` and compute the loss like this:
```
transformer = TransformerWrapper(...)
inp_seq, tgt_seq = sequence[..., :-1], sequence[..., 1:]
logits = transformer(inp_seq)
loss = torch.nn.functional.cross_entropy(
    logits.permute(0, 2, 1),
    tgt_seq,
    ignore_index=pad_idx,
)
```
which I believe is exactly equivalent to `AutoregressiveWrapper`, I am seeing a difference in the model convergence even though the starting loss at the first step of training is the same. With an `AutoregressiveWrapper` and computing the loss like this:
```
model = AutoregressiveWrapper(transformer)
loss = model(sequence)
```
the convergence is much faster (almost too good to be true). I cannot figure out from the code on why these two cases would be different. Am I missing anything?

Thank you.

**Comment by lucidrains** [comment]
@pradeep-pyro no that looks right, and if you [look at the code](https://github.com/lucidrains/x-transformers/blob/main/x_transformers/autoregressive_wrapper.py#L302-L326) it is the same

i'm not sure why convergence would be different

**Comment by lucidrains** [comment]
@pradeep-pyro could it be you have a few other transformer related features set on one script but not another? (say the value residual?)

**Comment by pradeep-pyro** [comment]
Thanks for the quick response (as always)! I am sure that the Transformer features are the same. The code looks like this in the model creation stage:

```
transformer_decoder = Decoder(
    dim=dim_emb,
    depth=depth,
    heads=heads,
    attn_dropout=attn_dropout,
    ff_dropout=ff_dropout,
    attn_flash=True,
    rotary_xpos=True,
)
transformer = TransformerWrapper(
    num_tokens=vocab_size,
    max_seq_len=512,
    use_abs_pos_emb=False,
    attn_layers=transformer_decoder,
    token_emb=embedder_layer,
    emb_dropout=emb_dropout,
)
self.pad_idx = token_to_index("PAD")
model = AutoregressiveWrapper(
    transformer,
    mask_prob=0,
    pad_value=self.pad_idx,
)
if self.transformer_impl == "transformer_wrapper":
    self.model = transformer
elif self.transformer_impl == "autoregressive_wrapper":
    self.model = model
```

Then for a single step (in PyTorch Lightning), I have
```
if self.transformer_impl == "transformer_wrapper":
    inp_seq, tgt_seq = sequence[..., :-1], sequence[..., 1:]
    logits = self.model(inp_seq)
    loss = torch.nn.functional.cross_entropy(
        logits.permute(0, 2, 1),
        tgt_seq,
        ignore_index=self.pad_idx,
    )
elif self.hparams.transformer_impl == "autoregressive_wrapper":
    loss = self.model(sequence)
```

I can try to make a fully reproducible script, I should probably do that anyway to isolate the problem.

**Comment by lucidrains** [comment]
hmm, I'm not sure!

**Comment by pradeep-pyro** [comment]
Hey @lucidrains, should the `ignore_index` here https://github.com/lucidrains/x-transformers/blob/main/x_transformers/autoregressive_wrapper.py#L325 be `pad_value` and not `ignore_index`. Looks like pad tokens are being considered in the loss.

**Comment by lucidrains** [comment]
oh I see, yeah you can just set the ignore_index on the wrapper to be the same as the pad value

**Comment by lucidrains** [comment]
mystery solved!

**Comment by pradeep-pyro** [comment]
Yes! Is this is a bug though? In this line https://github.com/lucidrains/x-transformers/blob/main/x_transformers/autoregressive_wrapper.py#L303 the ignore tokens are converted into pad tokens.

**Comment by lucidrains** [comment]
@pradeep-pyro no it isn't, because the `target` was set before that step

you also need to have natural numbers of indices or embedding step will break

**Comment by pradeep-pyro** [comment]
Ah yes, makes sense now. Thanks for the response and suggestions!
